proposals/xxxx-work-graphs.md

```diff
@@ -0,0 +1,545 @@
+<!-- {% raw %} -->
+
+# Work Graphs
+
+* Proposal: [NNNN](NNNN-work-graphs.md)
+* Author(s): [Chris Bieneman](https://github.com/llvm-beanz)
+* Sponsor: [Greg Roth](https://github.com/pow2clk), [Tex
+  Riddell](https://github.com/tex3d)
+* Status: **Under Consideration**
+* Planned Version: Shader Model 6.8
+
+## Introduction
+
+This document describes Work Graphs, a new feature for GPU-based work generation
+built on new Shader Model 6.8 DXIL features.
+
+Full documentation of the Work Graphs feature including DirectX runtime, HLSL,
+and DXIL documentation is available in the [Work Graphs Spec on
+GitHub](https://microsoft.github.io/DirectX-Specs/d3d/WorkGraphs.html). This
+document subsumes that document specifically as it relates to HLSL and the
+compiler features.
+
+## Motivation
+
+Current language and runtime limitations make generating work on GPU threads
+insufficient to meet the needs of some workloads. If existing GPU features (like
+ExecuteIndirect) can't sufficiently generate work the application must generate
+the work from the CPU resulting in unnecessary round-tripping between the GPU
+and CPU. Work Graphs solves this problem by enabling more robust GPU-based work
+creation APIs.
+
+## Proposed solution
+
+Work graphs allows an application to specify a set of tasks as _nodes_ in a
+graph representing a more complex workload. Each _node_ has a fixed shader which
+takes one or more _input records_ as input and can produce one or more _output
+records_ as output.
+
+### Launch Modes
+
+_Node_ shaders have one of three _launch modes_:
+
+* Thread
+* Broadcasting
+* Coalescing
+
+_Thread launch_ nodes represent an individual thread of work that processes a
+single _input record_. Thread launch nodes use a thread group size of `(1,1,1)`.
+
+_Broadcasting launch_ nodes represent a grid of work operating on a single
+_input record_. Each input record to a broadcasting node launches a full
+dispatch grid. The size of the dispatch grid can either be fixed for the node or
+specified in the input.
+
+_Coalescing launch_ nodes represent a thread group operating on a shared array
+of _input records_. The shader declares the maximum number of records per thread
+group.
+
+### Records
+
+_Input records_ represent inputs to node shaders, and _output records_ represent
+outputs from node shaders. Records can be singular or arrays. _Input records_
+can be read-only or read-write, while _output records_ are read-write but
+uninitialized when created.
+
+## Detailed design
+
+### Node Entry Functions
+
+Node shaders are compute shaders built into library targets. Shader entries
+annotated as `[shader("node")]` are usable as work graph nodes.
+
+A broadcasting launch mode node entry may also be a compute entry (annotated

```

Comment:
@pow2clk & @tex3d, I wonder whether or not we should change this. Do we get much benefit from having compute shaders work as node shaders? It wouldn't really be hard to wrap entry points to share shader bodies, and this seems to introduce some unnecessary complication in the language and compiler.